internal/kms/pubsub.go

```diff
@@ -0,0 +1,325 @@
+
+type PubSubServer interface {
+	AddPubSubTopic(string, rpc.MessageHandler) error
+	SendPubSubMessage(context.Context, string, []byte) (<-chan rpc.Response, error)
+}
+
+type pubSubService struct {
+	ctx             context.Context
+	peerID          libpeer.ID
+	pubsub          PubSubServer
+	keyRequestedSub *event.Subscription
+	eventBus        *event.Bus
+	encStore        *ipldEncStorage
+}
+
+var _ Service = (*pubSubService)(nil)
+
+func (s *pubSubService) GetKeys(ctx context.Context, cids ...cidlink.Link) (*encryption.Results, error) {
+	res, ch := encryption.NewResults()
+
+	err := s.requestEncryptionKeyFromPeers(ctx, cids, ch)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// NewPubSubService creates a new instance of the KMS service that is connected to the given PubSubServer,
+// event bus and encryption storage.
+//
+// The service will subscribe to the "encryption" topic on the PubSubServer and to the
+// "enc-keys-request" event on the event bus.
+func NewPubSubService(
+	ctx context.Context,
+	peerID libpeer.ID,
+	pubsub PubSubServer,
+	eventBus *event.Bus,
+	encstore datastore.Blockstore,
+) (*pubSubService, error) {
+	s := &pubSubService{
+		ctx:      ctx,
+		peerID:   peerID,
+		pubsub:   pubsub,
+		eventBus: eventBus,
+		encStore: newIPLDEncryptionStorage(encstore),
+	}
+	err := pubsub.AddPubSubTopic(pubsubTopic, s.handleRequestFromPeer)
+	if err != nil {
+		return nil, err
+	}
+	s.keyRequestedSub, err = eventBus.Subscribe(encryption.RequestKeysEventName)
+	if err != nil {
+		return nil, err
+	}
+	go s.handleKeyRequestedEvent()
+	return s, nil
+}
+
+func (s *pubSubService) handleKeyRequestedEvent() {
+	for {
+		msg, isOpen := <-s.keyRequestedSub.Message()
+		if !isOpen {
+			return
+		}
+
+		if keyReqEvent, ok := msg.Data.(encryption.RequestKeysEvent); ok {
+			go func() {
+				results, err := s.GetKeys(s.ctx, keyReqEvent.Keys...)
+				if err != nil {
+					log.ErrorContextE(s.ctx, "Failed to get encryption keys", err)
+				}
+
+				defer close(keyReqEvent.Resp)
+
+				select {
+				case <-s.ctx.Done():
+					return
+				case encResult := <-results.Get():
+					for _, encItem := range encResult.Items {
+						_, err = s.encStore.put(s.ctx, encItem.Block)
+						if err != nil {
+							log.ErrorContextE(s.ctx, "Failed to save encryption key", err)
+							return
+						}
+					}
+
+					keyReqEvent.Resp <- encResult
+				}
+			}()
+		} else {
+			log.ErrorContext(s.ctx, "Failed to cast event data to RequestKeysEvent")
+		}
+	}
+}
+
+// handleEncryptionMessage handles incoming FetchEncryptionKeyRequest messages from the pubsub network.
+func (s *pubSubService) handleRequestFromPeer(peerID libpeer.ID, topic string, msg []byte) ([]byte, error) {
+	req := new(pb.FetchEncryptionKeyRequest)
+	if err := proto.Unmarshal(msg, req); err != nil {
+		log.ErrorContextE(s.ctx, "Failed to unmarshal pubsub message %s", err)
+		return nil, err
+	}
+
+	ctx := grpcpeer.NewContext(s.ctx, net.NewGRPCPeer(peerID))
```

Comment:
todo: It doesn't make sense to create a dependency to the net package only for this. You can even create the `grpcpeer.Peer` struct inline if you want.

Reply:
done